### PR TITLE
Ruby: Revamp to make it up to date 

### PR DESCRIPTION
**Pull request by Andrenzo17**
- Implement Gopay tokenization
- Implement API subscription
- Add function Idempotency Key, override notification, append Notification
- Make new unit test using minitest
- Create example core api and snap using Sinatra framework
- Update the missing readme sections and remove Jquery usage from snap sample use pure JS sample
- Update version gemlock and bundler to satisfy dependabot alert

### CODE REVIEW

File: lib/test/api_test.rb [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+      card_exp_month: 12,
+      card_exp_year: 2025
+    }
+    get_token = @mt_test.test_token(card)
+    result = @mt_test.charge(
+      {
+        "payment_type": "credit_card",
+        "transaction_details": {
+          "gross_amount": 10000,
+          "order_id": "test-order-#{rand(1..10000)}"
```
Comment (rizdaprasetya) [inline lib/test/api_test.rb:41]: Becareful with `rand(1..10000` you may encounter order id duplicate after a while. It is better to use unix timestamp (better if the unix timestamp is accurate to exact milisecond), like this:
https://github.com/Midtrans/midtrans-nodejs-client/blob/b25bfd2e7364900893d515b24f3768c5fe1ea1c1/test/snap.test.js#L222

- if timestamp still produce duplicate order id, then add random digit suffix.
- Also make the order id prefix have clear indication it is coming from ruby lib test.

File: lib/test/api_test.rb [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        gross_amount: 200000
+      },
+      "credit_card": {
+        "secure": true
+      }
+    )
+    assert_equal 201, result.status_code
+  end
+
+end
```
Comment (rizdaprasetya) [inline lib/test/api_test.rb:259]: Any negative test case scenario? better add it, like this:
https://github.com/Midtrans/midtrans-nodejs-client/blob/master/test/coreApi.test.js#L78
search for `fail to`. Name your negative test case accordingly e.g: `test_fail_to_charge_with_empty_server_key`

File: lib/test/snap_test.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        order_id: "THE-ITEM-ORDER-#{rand(1..10000)}",
+        gross_amount: 200000
+      },
+      "credit_card": {
+        "secure": true
+      }
+    )
+    assert_equal 201, result.status_code
+  end
+
```
Comment (rizdaprasetya) [inline lib/test/snap_test.rb:28]: add some negative test case like when server key is empty or invalid. This is to ensure that your class throw/raise the correct error/exception

File: lib/test/gopay_tokenization_test.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    result = @mt_test.get_payment_account(charge.account_id)
+    assert_equal 201, result.status_code
+    assert_equal "PENDING", result.account_status
+  end
+
+  def test_unlink_payment_account
+    p 3
+    result = @mt_test.unlink_payment_account($test_gopay_id)
+    assert_equal "Account status cannot be updated.", result.status_message
+  end
```
Comment (rizdaprasetya) [inline lib/test/gopay_tokenization_test.rb:53]: is this expected that the test result are failing? if expected, rename it to e.g: `test_fail_to_unlink_...`

File: lib/test/transaction_test.rb [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  def test_refund_transaction
+    param = {
+      "refund_key": "reference1",
+      "amount": 5000,
+      "reason": "for some reason"
+    }
+    result = @mt_test.refund("dummy-order-id", param)
+    assert_equal 404, result.status_code
+    assert_equal "Transaction doesn't exist.", result.status_message
+  end
```
Comment (rizdaprasetya) [inline lib/test/transaction_test.rb:156]: this seem to be not a real positive case refund, better name it to indicate if this expected to fail or not. eg: test_fail_to_refund_unknown_transaction

File: lib/veritrans/api.rb [hunk 6]
```diff
@@ -59,6 +60,20 @@ def create_snap_token(options = {})
...
+      Veritrans::SnapResult.new(result.response, result.url, result.request_options, result.time)
+      result.token
+    end
+
+    # Create Snap URL payment
+    def get_snap_url(options = {})
+      result = request_with_logging(:post, config.api_host.sub('//api.', '//app.') + "/snap/v1/transactions", options)
+      Veritrans::SnapResult.new(result.response, result.url, result.request_options, result.time)
+      result.redirect_url
+    end
```
Comment (rizdaprasetya) [inline lib/veritrans/api.rb:75]: `get` is not the correct name, because `get` is only for retrieving. The function create & retrieve, so the name should be `create_`

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,10 @@
+### Version 2.4.0 (date 29 September 2021)
+* Implement Gopay tokenization
+* Implement API subscription
+* Add function Idempotency Key, override notification, append Notification
+* Make new unit test using minitest
+* Create example core api and snap using Sinatra framework
+
```
Comment (rizdaprasetya) [inline CHANGELOG.md:7]: add changelog about: update dependency, revamp implementation by removing non relevant features & adding new features

File: README.md [hunk 8]
```diff
@@ -6,247 +6,218 @@ Midtrans (formerly known as Veritrans Indonesia).
...
+    http_options: { }, # optional
+    logger: Logger.new(STDOUT), # optional
+    file_logger: Logger.new(STDOUT), # optional
+  )
+
+mt_client.status("order-id-123456")
+```
 
-### Create simple payment form (optional)
+also can set one by one
```
Comment (rizdaprasetya) [inline README.md:44]: wording should be more descriptive:
Alternatively, you can also set config by declaring each one like below:

File: README.md [hunk 9]
```diff
@@ -6,247 +6,218 @@ Midtrans (formerly known as Veritrans Indonesia).
...
+You can opt to add idempotency key on charge transaction. It can be achieved by adding additional HTTP headers into charge request.
+Is a unique value that is put on header on API request. Midtrans API accept Idempotency-Key on header to safely handle retry request
+without performing the same operation twice. This is helpful for cases where merchant didn't receive the response because of network issue or other unexpected error.
+```ruby
+$idempotency_key = "Unique-ID"
+```
+[More details](http://api-docs.midtrans.com/#idempotent-requests)
+
+
+## STEP 1: Process credit cards
```
Comment (rizdaprasetya) [inline README.md:77]: The old structure of this readme is no longer relevant (why is credit card step 1? why non card step 2? doesn't make sense), can you restructure like the other library we have update (e.g https://github.com/Midtrans/midtrans-nodejs-client#2-usage). So it is not step 1, step 2, but reader choose which product Snap, Core API, etc.

File: README.md [hunk 10]
```diff
@@ -359,19 +330,6 @@ Veritrans.file_logger = Logger.new("/my/important_logs/veritrans.log")
...
-
-    $ gem install veritrans
-
-**Usage**
-
-Testing http notification:
-
-    $ veritrans testhook http://localhost:3000/vt_events
-    $ veritrans testhook -o my-order-1 -c ~/path/to/veritrans.yml http://localhost:3000/vt_events
-
```
Comment (rizdaprasetya) [inline README.md:334]: is this `Veritrans::Events` section still relevant?

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
@Andrenzo17 Overall nice job 👍 thanks , with some feedbacks.

File: bin/midtrans [hunk 11]
```diff
@@ -1,3 +0,0 @@
-#!/usr/bin/env sh
-
-veritrans $@
```
Comment (rizdaprasetya) [inline bin/midtrans:3]: btw since bin files are deleted, when installing veritrans gem there is `warning` message

Screen Shot 2021-09-30 at 10 58 03

it can be fixed by removing this line: https://github.com/veritrans/veritrans-ruby/blob/809ea9869075f8de7a93d0ed82cdce6ee2dd122f/veritrans.gemspec#L18

File: example/sinatra/index.erb [hunk 12]
```diff
@@ -1,214 +1,189 @@
...
+                closeDialog();
+                alert(response.status_message);
+                $('button').removeAttr("disabled");
+            },
+            onPending: function(response){
+                console.log('pending');
+                console.log('response:',response);
+                closeDialog();
+            }
+        };
```
Comment (rizdaprasetya) [inline example/sinatra/index.erb:110]: The demo feels incomplete, why only output to console.log? can't seems to attempt another payment after success even. Can you update the example to be like the other library?
https://github.com/Xaxxis/midtrans-go/blob/e3b1ef0613036dc224ed59517e3aeb57d2f2edcb/example/simple/coreapi-card-3ds/views/index.html#L140-L156
or
https://github.com/Midtrans/midtrans-nodejs-client/blob/b25bfd2e7364900893d515b24f3768c5fe1ea1c1/examples/expressApp/views/simple_core_api_checkout.ejs#L121-L138

Comment (rizdaprasetya) [reply example/sinatra/index.erb:110]: also please rename the `sinatra.rb` into app.rb or webapp.rb, because sinatra is already on the folder name.

**Comment by Andrenzo17** [comment]
Kindly help to review again kak @rizdaprasetya 🙏
Thank you

File: example/sinatra/webapp.rb [hunk 13]
```diff
@@ -0,0 +1,77 @@
+require 'veritrans'
+require 'json'
+require 'sinatra'
+
+Veritrans.setup do
+  config.load_yml File.dirname(__FILE__) + "/veritrans.yml#development"
+end
+
+# Alternative way to to initialize core api client:
+# mt_client = Midtrans.new(
```
Comment (rizdaprasetya) [inline example/sinatra/webapp.rb:10]: Inconsistent: at above `Veritrans.setup` is used,  then below ` mt_client = Midtrans.new`. please be consistent.

File: example/sinatra/webapp.rb [hunk 14]
```diff
@@ -0,0 +1,77 @@
+require 'veritrans'
+require 'json'
+require 'sinatra'
+
+Veritrans.setup do
+  config.load_yml File.dirname(__FILE__) + "/veritrans.yml#development"
```
Comment (rizdaprasetya) [inline example/sinatra/webapp.rb:6]: Why are this demo using `config.load_yml File.dirname(__FILE__) + "/veritrans.yml#development"`, while the lib readme doesn't even mention this method of setup via YML file? you're confusing merchant by being inconsistent.

File: example/sinatra/webapp.rb [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+def generate_order_id
+  "sinatra-example-#{Time.now.to_i}"
+end
+
+get "/" do
+  erb :index
+end
+
+get "/snap" do
+  result = Veritrans.create_widget_token(
```
Comment (rizdaprasetya) [inline example/sinatra/webapp.rb:32]: why still use `create_widget_token` ? use the proper function named `snap`. You are confusing merchant.

File: example/sinatra/index.erb [hunk 16]
```diff
@@ -91,18 +91,21 @@
                 openDialog(redirect_url);
             },
             onSuccess: function(response){
+                document.getElementById('result-json').innerHTML += JSON.stringify(response, null, 2);
                 console.log('success');
                 console.log('response:',response);
                 closeDialog();
```
Comment (rizdaprasetya) [inline example/sinatra/index.erb]: are you sure this issue I told you earlier solved? Because I don't see `$('button').removeAttr("disabled");` here.
> can't seems to attempt another payment after success even. 

Are you copying this Core API HTML template from PHP? looks like it is a bad example i'll ask the maintainer to fix it too, you should copy from Go or NodeJS example

File: README.md [hunk 17]
```diff
@@ -149,7 +155,7 @@ result = Veritrans.create_widget_token(
 
 > !!! WARNING NOTE: VT-Web is deprecated, please use [Snap](#Snap Pop-up) instead, it has better previous VT-Web feature and many more improvements, including redirect_url.
 
-### VT-Direct / Core API
+###Core API (VT-Direct)
 
 It's little more complicated, because credit_card is sensitive data,
 you need put credit card number in our safe storage first using `midtrans.min.js` library, then send received token to with other payment details.
```
Comment (rizdaprasetya) [inline README.md:161]: Fix this, no longer `midtrans.min.js`

File: README.md [hunk 18]
```diff
@@ -129,7 +135,7 @@ Initialize Snap JS when customer click pay button
   </body>
 ```
 
-### Snap Redirect
+###Snap redirect
```
Comment (rizdaprasetya) [inline README.md]: You need to add spacing between `###` and the word. Look at this, it fail to render as proper heading: https://github.com/veritrans/veritrans-ruby/blob/441149af8176155e1ce667de1768b9c3cb93c1eb/README.md#2-choose-productmethod

File: README.md [hunk 19]
```diff
@@ -243,7 +247,7 @@ puts "Please transfer fund to account no. #{@result.permata_va_number} in bank P
 See [our documentation](https://api-docs.midtrans.com/#charge-features) for other available options.
 
 
-## STEP 3: Receive notification callback
+## Receive notification callback
```
Comment (rizdaprasetya) [inline README.md:250]: Why the other section have numbering, but this one have no number? be consistent please

File: README.md [hunk 20]
```diff
@@ -10,7 +10,7 @@ To see it in action, we have made example:
 
 1. Sinatra, which demonstrate in as succint code as possible. Please [have a look here](https://github.com/veritrans/veritrans-ruby/tree/master/example/sinatra)
```
Comment (rizdaprasetya) [inline README.md:11]: Screen Shot 2021-09-30 at 17 12 04
Readme structure is a mess, why is section 2 level is under 1? it should be on the same level.

So many word still using Veritrans, revamp it properly to Midtrans. Also still using vt-web, rename it to proper snap redirect, etc. I expect you to revamp the readme too, make it up to date with current situation, read them all and carefully revamp them all. Not just few part. Use other library's readme as a benchmark, especially NodeJS or Go, which is the most recently updated.

Change the overall structure and the content if needed, I expect you to make it up to date and clearer for merchant to understand it easily.

File: README.md [hunk 21]
```diff
@@ -10,7 +10,7 @@ To see it in action, we have made example:
...
 
-## How to use
+## 1. How to use
```
Comment (rizdaprasetya) [inline README.md:13]: `Usage` is better than `How to use`

File: README.md [hunk 22]
```diff
@@ -149,7 +155,7 @@ result = Veritrans.create_widget_token(
 
 > !!! WARNING NOTE: VT-Web is deprecated, please use [Snap](#Snap Pop-up) instead, it has better previous VT-Web feature and many more improvements, including redirect_url.
```
Comment (rizdaprasetya) [inline README.md]: no longer relevant

File: README.md (hunk 17 above)
Comment (rizdaprasetya) [reply README.md:161]: also improve the wording `It's little more complicated, because credit_card is sensitive data` is not a good one.

File: README.md (hunk 20 above)
Comment (rizdaprasetya) [reply README.md:11]: I expect you to think & take initiative in improving the readme, not just waiting for me to give order which part to change, no one want to be robot and just take orders.

Comment (rizdaprasetya) [reply README.md:11]: This library readme should be at least as good as Go or NodeJS library, or better. That's the standard. Currently the readme is a mess and outdated.

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
There's a lot to improve especially on readme, please address. Thanks @Andrenzo17

File: README.md (hunk 20 above)
Comment (rizdaprasetya) [reply README.md:11]: The Core API part is also a mess, the sample HTML is not enough for card transaction, there is no guide (or link) on how to open 3DS, handle the JS callback etc.

File: example/sinatra/index.erb (hunk 16 above)
Comment (Andrenzo17) [reply example/sinatra/index.erb]: Yes kak i'm copying from PHP example. Thank you for your direction

File: README.md (hunk 20 above)
Comment (Andrenzo17) [reply README.md:11]: i'm so sorry kak, I'm working in a hurry because I think I have to finish the revision today. In the future I will be more careful and not in a hurry
Thanks a lot ka

**Comment by Andrenzo17** [comment]
Hi kak @rizdaprasetya sorry for all the revision, once again kindly help to review 🙏
Thank you so much

File: README.md [hunk 23]
```diff
@@ -1,345 +1,245 @@
-# Veritrans ruby library
+# Midtrans Ruby library
```
Comment (rizdaprasetya) [inline README.md:2]: Write guide to maintaining this repo e.g. https://github.com/Midtrans/midtrans-nodejs-client/blob/master/Maintaining.md, at least how to run the test etc for this repo if our other team member want to maintain this package.

Comment (rizdaprasetya) [reply README.md:2]: No need how-to update ruby gem yet if you don't know, since we haven't got access.

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
Thanks @Andrenzo17 for the progress, still a lot to improve actually. Please proceed. thanks!

File: README.md [hunk 24]
```diff
@@ -1,345 +1,245 @@
...
-    # assuming you have model Order in your project
-    Order.find_by(order_id: payment.order_id) ...
-  end
-end
+$idempotency_key = "Unique-ID"
 ```
+[More details](http://api-docs.midtrans.com/#idempotent-requests)
 
-#### Logging
+### Logging
```
[the remark this answers is not in the document]
Comment (Andrenzo17) [reply README.md:243]: This code for Create log folder if not exists, for Rails user will write log to RAILS_ROOT/log/veritrans.log kak

[the remark this answers is not in the document]
Comment (rizdaprasetya) [reply README.md:243]: @Andrenzo17 Yes I know, but that's not the question. Can you answer my original question or do I need to clarify it?

File: lib/veritrans.rb [hunk 25]
```diff
@@ -4,24 +4,18 @@
 require 'veritrans/api'
 require 'veritrans/result'
 
-if defined?(::Rails)
-  require 'veritrans/events'
-end
```
[the remark this answers is not in the document]
Comment (rizdaprasetya) [reply lib/veritrans.rb:9]: deleting CLI is fine since it may not directly used in the code, but deleting Events is not.

File: lib/veritrans.rb [hunk 26]
```diff
@@ -183,13 +177,7 @@ def file_logger
     if !@file_logger
       require 'logger'
       begin
-        if defined?(Rails) && Rails.root
-          require 'fileutils'
-          FileUtils.mkdir_p(Rails.root.join("log"))
-          @file_logger = Logger.new(Rails.root.join("log/veritrans.log").to_s)
-        else
           @file_logger = Logger.new("/dev/null")
-        end
```
Comment (rizdaprasetya) [inline lib/veritrans.rb:193]: I was not asking you to remove it actually. I was asking you to test if the behavior will be as expected or not on rails, will explain via slack

File: api_reference.md [hunk 27]
```diff
@@ -121,29 +126,14 @@ our [documentation](https://api-docs.midtrans.com/)
...
+## Usage Example
 
-q.class # => Midtrans::Result
-q.data == {
-  status_code: "201",
-  "token": "2b3ccb6c-d0fb-499a-9d46-ef53ad51fe62",
-  "redirect_url": "https://app.sandbox.midtrans.com/snap/v2/vtweb/2b3ccb6c-d0fb-499a-9d46-ef53ad51fe62"
-}
-```
+### Create Transaction
```
Comment (rizdaprasetya) [inline api_reference.md:131]: Why is every sub-section under this `Create Transaction` section level now? Only Core API & Snap create transaction should be under this right?

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
There's good progress, but still need few things to improve. Thanks @Andrenzo17

File: example/snap/snap_example.rb [hunk 28]
```diff
@@ -24,6 +24,8 @@
   }
 )
 
+puts result.body
```
Comment (rizdaprasetya) [inline example/snap/snap_example.rb:27]: why printing the `.body`? should print `.data` instead, because in real use case JSON response attribute access is through the `.data[:attribute_name]` right?

File: api_reference.md [hunk 29]
```diff
@@ -586,9 +607,9 @@ param = {
   }
 }
 
-q = Midtrans.update_subscription("d137e7f4-9474-4fc2-9847-672e09cb16f6", param)
+charge_response = Midtrans.update_subscription("d137e7f4-9474-4fc2-9847-672e09cb16f6", param)
 
-q == {
+charge_response.body == {
```
Comment (rizdaprasetya) [inline api_reference.md:612]: why all these example is using `.body`? real case usage will be through `.data`, so maybe change the example format to

```
puts result.data # this will be Hash representation of the API JSON response, of example:
# <paste the json here>
```

File: api_reference.md [hunk 30]
```diff
@@ -605,6 +626,6 @@ result.class # => Midtrans::Result
 * `Midtrans::Result#status_code` - `integer`, e.g. 200, 402. Documentation https://api-docs.midtrans.com/#status-code
```
Comment (rizdaprasetya) [inline api_reference.md:632]: Why is this section under level 4 heading? it should be level 2 heading for proper section leveling.

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
@Andrenzo17 Ok good progress, still need more improvement though. Thanks

Also more suggestion:
1. example readme command seems to fail on my machine, should update it to:
```
2. Run the sinatra app: `bundle exec ruby webapp.rb`
```
To ensure ruby will run the script with gem installed by bundle

2. Example folder of Snap & Core API should also have readme.md file, at least tell how to install dependency. then run the script. Because the script is not directly runable without installing dependency from terminal. Or find a way to make the script require dependency from local folder of this lib

File: api_reference.md [hunk 31]
```diff
@@ -607,19 +622,26 @@ param = {
...
 
-#### `Midtrans::Result`
-
+## Get json field value from Midtrans API result
 ```ruby
 result = Midtrans.charge(...)
 
-result.class # => Midtrans::Result
+# If in the future there is a new attribute in the Midtrans Result.
+# Recommended retrieve like example below.
```
Comment (rizdaprasetya) [inline api_reference.md:638]: please mind the grammar & wording. Should be something like:
> The recommended way to retrieve the value is like example below.

File: api_reference.md [hunk 32]
```diff
@@ -607,19 +622,26 @@ param = {
...
+# :new_attribute is the new field name in API response JSON
+result.data[:new_attribute]
+
+# Alternatively you can also access like below.
```
Comment (rizdaprasetya) [inline api_reference.md:642]: In previous comment you mention `retrieve`, but here you mention `access`, choose one and be consistent with the wording. Consistency is important when writing documentation.

File: api_reference.md [hunk 33]
```diff
@@ -607,19 +622,26 @@ param = {
...
+# But make sure the new field name already show in API response JSON, if not will trigger error.
```
Comment (rizdaprasetya) [inline api_reference.md:643]: Grammar
> But please make sure the new field name is actually exists on the API response JSON, else it will trigger an error.

or better wording
> Note: When the field name is not actually exists on the API response JSON, it will trigger an error. Make sure to handle it.

Comment (rizdaprasetya) [reply api_reference.md:643]: Which remind me of: you don't even document handling error / exception. Didn't I tell you to refer to the NodeJS repo, which it have that section right? Why not adding it on this repo? At least demonstrate what kind of attributes is accessible from the error object

File: example/coreapi/core_api_credit_card_example.rb [hunk 34]
```diff
@@ -37,23 +37,24 @@
   }
 )
 
-puts result.body
+puts result.data
```
Comment (rizdaprasetya) [inline example/coreapi/core_api_credit_card_example.rb:40]: - Add example of handling/catching error
- Also don't just plainly print out variable. Add some descriptive text, e.g. print out "Charge result: "

Both of above are exist on the nodejs example as the reference, the example already have that, so that is the standard-quality, which means this example have less-quality than the reference. We should upgrade the quality of our work (learning from our past work), not downgrade.

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
Still few things to improve, thanks @Andrenzo17 !

**Comment by Andrenzo17** [comment]
Thanks again for review and feedback kak.
Kindly help to review @rizdaprasetya 
thank you so much

File: lib/veritrans/midtrans_error.rb [hunk 35]
```diff
@@ -0,0 +1,10 @@
+class MidtransError < StandardError
+  attr_reader :status
+  attr_reader :data
+
+  def initialize(message, status_code = nil, api_response = nil)
+    super(message)
+    @status = status_code
+    @data = api_response
+  end
```
Comment (rizdaprasetya) [inline lib/veritrans/midtrans_error.rb:9]: As previously reminded, please mind these points:

### Descriptive var/attribute names:
`status` is not very descriptive, what status do you mean: payment? API? look at the [example](https://github.com/Midtrans/midtrans-nodejs-client#3-handling-error--exception), should be something like `http status code`.

`data` what data do you even mean?

### Your implementation should be at least on same standard or better than the example/reference:
The reference have proper descriptive attributes name, your implementation is slightly worse. Why your implementation don't include the `rawHttpClientData` ? Please add, the name can also improved to be more descriptive to `raw http client response`

File: lib/veritrans/client.rb [hunk 36]
```diff
@@ -74,7 +74,7 @@ def patch(url, params)
 
     def make_request(method, url, params, auth_header = nil)
       if !config.server_key || config.server_key == ''
-        raise "Please add server_key to config/veritrans.yml"
+        raise "Please config your server key"
```
Comment (rizdaprasetya) [inline lib/veritrans/client.rb:77]: Use formal language for error message, should be `configure your server key`

File: example/coreapi/core_api_credit_card_example.rb [hunk 37]
```diff
@@ -15,29 +15,34 @@
...
+      "credit_card": {
+        "token_id": "#{get_token.token_id}"
+      }
+    }
+  )
+  puts "charge result : #{result.data}"
+rescue MidtransError => e
+  puts e.message # Basic message error
+  puts e.status # HTTP status code e.g: 400, 401, etc.
+  puts e.data # JSON of the API response
```
Comment (rizdaprasetya) [inline example/coreapi/core_api_credit_card_example.rb:44]: It is not JSON in this case right? Shouldn't it be `API response body in String` ?

File: README.md [hunk 38]
```diff
@@ -273,7 +273,21 @@ end
...
+When using function that result in Midtrans API call e.g: `Midtrans.charge(...)` or `Midtrans.create_snap_token(...)`
+there's a chance it may throw error (`MidtransError` object), the error object will contains below properties that can be used as information to your error handling logic:
+
+```ruby
+begin
+  Midtrans.create_snap_token(parameter)
+rescue MidtransError => e
+  puts e.message # Basic error message string
+  puts e.status # HTTP status code e.g: 400, 401, etc.
+  puts e.data # JSON of the API response
```
Comment (rizdaprasetya) [inline README.md:286]: It is not JSON in this case right? Shouldn't it be `API response body in String` ?

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
Looks better, but still few things to improve. Thanks.

Also since we are introducing changes that may not backward compatible, in Readme please add something like this
```
## Important Changes

### v2.4.0
- API client methods will now raise `MidtransError` when getting unexpected API response. You may need to update your error handling. <link to the handling error section of readme >
- Removed features: <mention what is removed, like the CLI, etc> . Mainly removed due to no longer relevant/essential to this library's purpose.
```

**Comment by Andrenzo17** [comment]
Done for request changes kak.
Kindly help to review @rizdaprasetya 
Thank you very much

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
Still few things to fix. thanks @Andrenzo17

**Review by rizdaprasetya — APPROVED** [review]
Looks good now 🎉 thanks @Andrenzo17 ! Approved to merge & release to ruby gem, for Monday.

**Comment by rizqirizqi** [comment]
@rizdaprasetya @Andrenzo17 Sorry to come to this PR, because a working feature suddenly is not working on my side (The Veritrans CLI) and turns out you guys expected the backward incompatibility. This change should've been a major version update. 😅
